### PR TITLE
Temporary privileges as a service

### PR DESCRIPTION
**Pull request by JPLachance**
# What

In the past years, we worked a lot on improving how employees can get access to environments securely, temporarily. My team spent a lot of time on this project, strongDM found it super cool, they submitted our project to Identity Defined Security Alliance, we won the [Identity Management Project of the Year – SMB](https://www.idsalliance.org/press-release/identity-defined-security-alliance-announces-winners-of-identity-management-awards/) award.

# Goals

Talk about the project, promote [our job opening](https://www.coveo.com/en/company/careers/open-positions/research-and-development/senior-developer-security/6616293002).

# Rendered

https://github.com/JPLachance/source.coveo.com/blob/temporary-privileges-system/_posts/2023-04-11-temporary-privileges.md

### CODE REVIEW

File: _posts/2023-04-11-temporary-privileges.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+  bio: Staff Software Developer, Team Lead, R&D Security Defense
+  twitter: JeePLachance
+  image: jplachance.jpg
+---
+
+I spend most of my days inside a code editor, Jira, Outlook, or in meetings. Now that one of our projects won an award (link to come), it's a great opportunity to take a look back at what the Coveo R&D Defense team built.
+
+The Coveo infrastructure is constantly growing. DevOps engineers add new regions, new services, which leads to more systems that can break, more complex access management and more complex audit logging. If I tell stakeholders that the entire R&D department needs always-on access to all the services they deploy and own in a production environment, some of those stakeholders will tell me that the risks are too high, that it is not acceptable. On the other hand, if only a handful of people can help when there is an incident in production, if someone needs to wake up the on-call access management person at 3 in the morning because he needs access to a specific role in a specific database in our Australian region, the time to resolution of incidents goes up, our service level agreement gets breached, leadership does not like that.
```
Comment (lbourdages) [inline _posts/2023-04-11-temporary-privileges.md:17]: `because he needs`-> `because they need`

File: _posts/2023-04-11-temporary-privileges.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+This is why Coveo needed a good middle ground. The R&D department needed a system that allows selected employees to gain privileged access on systems they own for a short period of time, fix the incident, follow up with a post-mortem. Back in 2020, Coveo adopted [strongDM](https://www.strongdm.com/) to manage privileged access rights. While it already supported granting temporary privileges, it lacked a way to allow employees to quickly request a temporary privileges, without waking up the strongDM administrator at 3 AM. From the strongDM APIs, the R&D Defense team built that system.
+
+<!-- more -->
+
+
+## Tracking requirements
+
+What do a Coveo employee needs before being allowed to request a temporary privilege to a system the employee owns? Well, many things. The first step is to complete the Coveo onboarding process. This process will make the employee read Coveo policies, go through trainings, go through a background check procedure. Coveo has what HR calls the Coveo 90. The employee manager is responsible for going through the entire process with the new employee.
```
Comment (ppferland) [inline _posts/2023-04-11-temporary-privileges.md]:
```suggestion
What does a Coveo employee need before being allowed to request a temporary privilege to a system the employee owns? Well, many things. The first step is to complete the Coveo onboarding process. This process will make the employee read Coveo policies, go through trainings, go through a background check procedure. Coveo has what HR calls the Coveo 90. The employee manager is responsible for going through the entire process with the new employee.
```

File: _posts/2023-04-11-temporary-privileges.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+          - system: STRONGDM
+            sub_system: clusters
+            privileges:
+              - prod-platform-activity-service-maintenance
+          - system: IAM
+            privileges:
+              - user-group-prod-temp-platform-admin
+```
+
+Some environments have `requirements`, some don't. For example, the development environment is one where Coveo does enforce strict requirements, to allow quick frictionless innovation. The production environment though is a lot more restricted.
```
Comment (lbourdages) [inline _posts/2023-04-11-temporary-privileges.md]: `the development environment is one where Coveo does enforce strict requirements`

Does not?

File: _posts/2023-04-11-temporary-privileges.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+```YAML
+requirements:
+  background_check: true
+  supervisor_approval: true
+  trainings:
+    - hipaa-training
+    - policy-acknowledgement
+```
+
+In this specific example, the access matrix defines that a background check be completed, that the employee supervisor endorsement is required, that some specific trainings in our Learning Management System must be completed.
```
Comment (lbourdages) [inline _posts/2023-04-11-temporary-privileges.md]: *and* that some specific trainings [...]

File: _posts/2023-04-11-temporary-privileges.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+## Learning Management System (LMS) integration
+
+Off course, the access management system then needs to be fully integrated with a few system, like the LMS. Coveo invested a lot in learning management by creating [level up](https://levelup.coveo.com/), a home-made LMS, tailor made to fit our needs. On level up, Coveo has internal and external trainings. Anyone can learn how to use the Coveo Platform through level up. On top of that, employees can learn about internal policies and procedures as well.
+
+With this building block, each employee has a clear path on level up. After completing a training, the access management system picks it up, then the employee can automatically start to see more resources in the resources catalog. More on that later!
+
+## Power to the managers
+
+One of the main requirements is the manager approval. Traditionally, a manager opens a request to an IT team to request someone from IT to add his new hire in a given group. The other option is that an IT person checks the new hire colleague groups, applies a copy-paste logic. In both cases, there is either inefficiencies or over provisioning of privileges.
```
[the remark this answers is not in the document]
Comment (lbourdages) [reply _posts/2023-04-11-temporary-privileges.md:97]: *and* applies

[the remark this answers is not in the document]
Comment (lbourdages) [reply _posts/2023-04-11-temporary-privileges.md:97]: there is either -> there are either

File: _posts/2023-04-11-temporary-privileges.md (hunk 2 above)
Comment (gruel-coveo) [inline _posts/2023-04-11-temporary-privileges.md:26]: _What do a Coveo employee needs_ -> **What does a Coveo employee need**

File: _posts/2023-04-11-temporary-privileges.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+This is why, on top of our temporary privileges solution, the R&D Defense team built a small API and frontend application to allow a manager to grant up to his/her own privileges to a new hire, once all the requirements for that specific team and level are met.
```
Comment (lbourdages) [inline _posts/2023-04-11-temporary-privileges.md:99]: his/her -> their

File: _posts/2023-04-11-temporary-privileges.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+![]({{ site.baseurl }}/images/2023-04-11-temporary-privileges/uam_privileges_edit.png)
+
+A Coveo manager:
+
+- Can see the employee progression in the LMS
+- Can select the employee access group and access level, up to his own privileges
```
Comment (lbourdages) [inline _posts/2023-04-11-temporary-privileges.md:106]: his -> their

File: _posts/2023-04-11-temporary-privileges.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+  trainings:
+    - hipaa-training
+    - policy-acknowledgement
+```
+
+In this specific example, the access matrix defines that a background check be completed, that the employee supervisor endorsement is required, that some specific trainings in our Learning Management System must be completed.
+
+## Learning Management System (LMS) integration
+
+Off course, the access management system then needs to be fully integrated with a few system, like the LMS. Coveo invested a lot in learning management by creating [level up](https://levelup.coveo.com/), a home-made LMS, tailor made to fit our needs. On level up, Coveo has internal and external trainings. Anyone can learn how to use the Coveo Platform through level up. On top of that, employees can learn about internal policies and procedures as well.
```
Comment (ppferland) [inline _posts/2023-04-11-temporary-privileges.md:91]:
```suggestion
Of course, the access management system then needs to be fully integrated with a few system, like the LMS. Coveo invested a lot in learning management by creating [level up](https://levelup.coveo.com/), a home-made LMS, tailor-made for our needs. On level up, Coveo has internal and external trainings. Anyone can learn how to use the Coveo Platform through level up. On top of that, employees can learn about internal policies and procedures as well.
```

File: _posts/2023-04-11-temporary-privileges.md [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+- Privileges granted to employees
+- Etc.
+
+The interface is built using React and [Coveo Plasma](https://plasma.coveo.com/), our library of React components built on top of [Mantine](https://mantine.dev/).
+
+The access matrix is stored in GitHub. Any change to the access matrix goes through a pull request, a code review, approval by managers and asset owners. The beauty of having this matrix in the YAML format in GitHub is that any employee, manager, asset owner can contribute to it, suggest changes, improve the life of Coveo employees. Once approved, changes are synced into AWS S3, the Lambda function picks it up, the available resources catalog is updated. This is a classic [GitOps](https://about.gitlab.com/topics/gitops/) example.
+
+## The post-mortem
+
+Some peers in the security space will say that allowing an engineer to get that access with such low friction is unsafe. I see the point, but the layers of multi-factor authentications, VPNs, all the logging made by our API, all the logging made by strongDM, combine with all I described above allows me to sleep well at night.
```
Comment (lbourdages) [inline _posts/2023-04-11-temporary-privileges.md:147]: combine -> combined

File: _posts/2023-04-11-temporary-privileges.md [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
```
Comment (ppferland) [inline _posts/2023-04-11-temporary-privileges.md:147]:
```suggestion
Some peers in the security space will say that allowing an engineer to get that access with such low friction is unsafe. I see the point, but the layers of multi-factor authentications, VPNs, all the logging made by our API, all the logging made by strongDM, combined with all I described above allows me to sleep well at night.
```

**Review by ppferland — APPROVED** [review]

File: _posts/2023-04-11-temporary-privileges.md (hunk 1 above)
Comment (lbourdages) [reply _posts/2023-04-11-temporary-privileges.md]: On the other hand, if only a handful of people can help when there is an incident in production, the on-call access management person will have to be woken up every time a dev needs access to a specific resource. This makes access management unhappy, as well as increases the time to resolution and could even cause a breach of our service level agreement. Leadership won't like that.

**Review by cyb3rjerry — APPROVED** [review]

File: _posts/2023-04-11-temporary-privileges.md [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+```YAML
+requirements:
+  background_check: true
+  supervisor_approval: true
+  trainings:
+    - hipaa-training
+    - policy-acknowledgement
+```
+
+In this specific example, the access matrix defines that a background check be completed, that the employee's supervisor endorsement is required, and that some specific trainings in our Learning Management System must be completed.
```
Comment (JPLachance) [inline _posts/2023-04-11-temporary-privileges.md:87]:
```suggestion
In this specific example, the access matrix defines that a background check must be completed, that the employee's supervisor endorsement is required, and that some specific trainings in our Learning Management System must be completed.
```

**Review by lbourdages — APPROVED** [review]

**Review by tbronsonCoveo — APPROVED** [review]

File: _posts/2023-04-11-temporary-privileges.md [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+This is why Coveo needed a good middle ground. The R&D department needed a system that allows selected employees to gain privileged access on systems they own for a short period of time, fix the incident, follow up with a post-mortem. Back in 2020, Coveo adopted [strongDM](https://www.strongdm.com/) to manage privileged access rights. While it already supported granting temporary privileges, it lacked a way to allow employees to quickly request a temporary privilege, without waking up the strongDM administrator at 3 AM. From the strongDM APIs, the R&D Defense team built that system.
```
Comment (fverpaelst) [inline _posts/2023-04-11-temporary-privileges.md:19]: I suggest "[...] then follow up with a post-mortem".

File: _posts/2023-04-11-temporary-privileges.md [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+
+<!-- more -->
+
+
+## Tracking requirements
+
+What does a Coveo employee need before being allowed to request a temporary privilege to a resource in our infrastructure the employee owns? Well, many things. The first step is to complete the Coveo onboarding process. This process will make sure the employee read the Coveo policies, went through trainings, cleared the background check, etc. Coveo has what HR calls the Coveo 90. The employee manager is responsible for going through the entire process with the new employee.
+
+So what's the link with temporary privileges? Well, the access management system must not grant a privilege to an employee that has not met all these requirements yet. How does Coveo define the requirements? In an access matrix. Ok, what? Well, Coveo R&D employees are invited to contribute to a centralized YAML file that defines:
```
Comment (fverpaelst) [inline _posts/2023-04-11-temporary-privileges.md:28]: Paragraph feels iffy... Suggest re-write. Suggestion: "The activities a new employee undergoes through the Coveo 90 fulfills a set of requirements that only then allows access management to grant certain privileges to that employee. Each team might have different access requirements, which are defined in an access matrix, consisting in a centralized file that defines:"

**Review by fverpaelst — COMMENTED** [review]
Put 2 comments in, excellent post!

**Pull request merged**